### ISSUE TITLE
[question] Can't build CMake project via conan, but using Visual Studio works flawlessly

### ISSUE THREAD

**Issue by famastefano** [conan-io/conan#12791]
### What is your question?

Hello,
I have a project that is being ported from vcpkg.
I'm using Visual Studio + CMake as it is supported natively.
Install the requirements with conan works.
Configuring and building the project _with Visual Studio_, works, but not with conan.
Interestingly, it seems that the configuration works, but not the building phase, because if I manually open the .sln file, and build via Visual Studio, it works!
I don't understand what's going on :|

```
PS C:\Users\fama\source\repos\dgshape-controller> conan create . -pr:h=win64_dbg -pr:b=win64_dbg --build=missing
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute
Exporting package recipe
dgshape-controller/3.0.0 exports_sources: Copied 8 '.dll' files
dgshape-controller/3.0.0 exports_sources: Copied 3 '.txt' files: CMakeLists.txt, CMakeLists.txt, CMakeLists.txt
dgshape-controller/3.0.0 exports_sources: Copied 2 '.json' files: CMakePresets.json, CMakePresets.json
dgshape-controller/3.0.0 exports_sources: Copied 3 '.cmake' files: install-config.cmake, install-rules.cmake, variables.cmake
dgshape-controller/3.0.0 exports_sources: Copied 17 '.hpp' files
dgshape-controller/3.0.0 exports_sources: Copied 1 '.inc' file: model_traits.hpp.inc
dgshape-controller/3.0.0 exports_sources: Copied 4 '.h' files: DWX52DControlAPI.h, DWX52DCControlAPI.h, DWX52DCiControlAPI.h, DWX53DCControlAPI.h
dgshape-controller/3.0.0 exports_sources: Copied 20 '.cpp' files
dgshape-controller/3.0.0: A new conanfile.py version was exported
dgshape-controller/3.0.0: Folder: C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\export
dgshape-controller/3.0.0: Using the exported files summary hash as the recipe revision: 5ae3bfd2d7b714b7844c99dc0b150751
dgshape-controller/3.0.0: Package recipe modified in export, forcing source folder removal
dgshape-controller/3.0.0: Use the --keep-source, -k option to skip it
dgshape-controller/3.0.0: Exported revision: 5ae3bfd2d7b714b7844c99dc0b150751
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=17
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=17
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

dgshape-controller/3.0.0: WARN: Package binary is corrupted, removing: 296bf20223654528e469e28205f5cbbfb4ba26be
Version ranges solved
    Version range '>=1.10.0' required by 'dgshape-controller/3.0.0' resolved to 'spdlog/1.11.0' in local cache

Installing package: dgshape-controller/3.0.0
Requirements
    dgshape-controller/3.0.0 from local cache - Cache
    fmt/9.1.0 from 'conancenter' - Cache
    spdlog/1.11.0 from 'conancenter' - Cache
Packages
    dgshape-controller/3.0.0:296bf20223654528e469e28205f5cbbfb4ba26be - Build
    fmt/9.1.0:2c52a23dc25833dd4323e8144302393f0061b96b - Cache
    spdlog/1.11.0:80ba5fa094e2c380295741ebe5ab58dc37ba6887 - Cache

Installing (downloading, building) binaries...
fmt/9.1.0: Already installed!
spdlog/1.11.0: Already installed!
dgshape-controller/3.0.0: WARN: Build folder is dirty, removing it: C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\build\296bf20223654528e469e28205f5cbbfb4ba26be
dgshape-controller/3.0.0: Configuring sources in C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\source\.
dgshape-controller/3.0.0: Copying sources to build folder
dgshape-controller/3.0.0: Building your package in C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\build\296bf20223654528e469e28205f5cbbfb4ba26be
dgshape-controller/3.0.0: Generator 'CMakeDeps' calling 'generate()'
dgshape-controller/3.0.0: Generator 'CMakeToolchain' calling 'generate()'
dgshape-controller/3.0.0: Preset 'default' added to CMakePresets.json. Invoke it manually using 'cmake --preset default'
dgshape-controller/3.0.0: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Visual Studio 17 2022" -DCMAKE_TOOLCHAIN_FILE=C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\build\296bf20223654528e469e28205f5cbbfb4ba26be\build\generators\conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW'
dgshape-controller/3.0.0: Aggregating env generators
dgshape-controller/3.0.0: Calling build()
Preset CMake variables:

  CMAKE_BUILD_TYPE="Debug"
  CMAKE_CXX_FLAGS="/utf-8 /W4 /permissive- /volatile:iso /Zc:preprocessor /EHsc /Zc:__cplusplus /Zc:externConstexpr /Zc:throwingNew"
  CMAKE_TOOLCHAIN_FILE="C:/Users/fama/.conan/data/dgshape-controller/3.0.0/_/_/build/296bf20223654528e469e28205f5cbbfb4ba26be/build/generators/conan_toolchain.cmake"
  DGSHAPE_CONTROLLER_BUILD_TESTS="ON"
  DGSHAPE_CONTROLLER_DEBUG_BREAK="ON"

-- Using Conan toolchain: C:/Users/fama/.conan/data/dgshape-controller/3.0.0/_/_/build/296bf20223654528e469e28205f5cbbfb4ba26be/build/generators/conan_toolchain.cmake
-- Conan toolchain: Setting BUILD_SHARED_LIBS = ON
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
-- The CXX compiler identification is MSVC 19.32.31335.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.32.31326/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Component target declared 'spdlog::spdlog'
-- Conan: Component target declared 'fmt::fmt'
-- Compiling as x64
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Failed
-- Performing Test COMPILER_HAS_DEPRECATED
-- Performing Test COMPILER_HAS_DEPRECATED - Success
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_INSTALL_DATAROOTDIR
    CONAN_COMPILER
    CONAN_COMPILER_VERSION
    CONAN_EXPORTED
    CONAN_IN_LOCAL_CACHE
    CONAN_LINK_RUNTIME


-- Build files have been written to: C:/Users/fama/.conan/data/dgshape-controller/3.0.0/_/_/build/296bf20223654528e469e28205f5cbbfb4ba26be/build/win64-dbg
Error: could not load cache
dgshape-controller/3.0.0:
dgshape-controller/3.0.0: ERROR: Package '296bf20223654528e469e28205f5cbbfb4ba26be' build failed
dgshape-controller/3.0.0: WARN: Build folder C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\build\296bf20223654528e469e28205f5cbbfb4ba26be\build
ERROR: dgshape-controller/3.0.0: Error in build() method, line 75
        cmake.build()
        ConanException: Error 1 while executing cmake --build C:\Users\fama\.conan\data\dgshape-controller\3.0.0\_\_\build\296bf20223654528e469e28205f5cbbfb4ba26be\build --config Debug -- /m:8 /verbosity:minimal
```

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by famastefano** [comment]
Ah, fixed it thanks to the conan documentation being great (so fare at least).

Inside my `conanfile.py` I'm doing something like this:
```python
def build(self):
        cmake = CMake(self)
        preset = ""
        if(self.settings.os == "Windows"):
            preset="something"
        
        cmake.configure(args=[f"--preset={preset}"])
        cmake.build()
```
But my presets actually generate the configuration inside `${sourceDir}/build/${presetName}`, thus giving me the error because:
1. The configuration is generated into `whatever-pkg/build/my-preset`
2. CMake tries to build `whatever-pkg/build`, that doesn't have any cache as it's not where the config is.

I've fixed it by simply specifying the build dir:
```python
def build(self):
        cmake = CMake(self)
        preset = ""
        if(self.settings.os == "Windows"):
            preset="something"
        
        cmake.configure(args=[f"--preset={preset}"])
        cmake.build(build_dir=f"{cmake.build_folder}/{preset}")
```

**Comment by memsharded** [comment]
Hi @famastefano 

Glad that you found the issue and managed to make it work.
Note, however, that in theory the ``CMake`` helper is able to get the information from the Conan generated presets, and apply it to the ``cmake`` invocation, via arguments like ``-DXXX=YYY``.  Conan should be able to map the settings, and resolve to the correct presets file, and apply it to the build. If this is not happening, please feel free to re-open and provide details, it would be great to fully understand what could be happening if not working as we expected.

**Comment by famastefano** [comment]
Hi @memsharded,
as always the culprit were me.
First of all the presets are no longer needed as I can handle CMake options inside conan, so I've removed them completely.
Second, I was using the old CMake class from `conans`, because the one from `conan.tools.cmake` doesn't have the `definitions` attribute. But I handled that too by constructing a dictionary and passing it to the `configure` method.

`conans.CMake` doesn't pass the toolchain file, so the dependences weren't found.
`conan.tools.cmake.CMake` works flawlessly.
